### PR TITLE
refactor CreateStateDir

### PR DESCRIPTION
**Pull request by kalensk**
Simplify logic/readability using upgrade.PathExist. Clean up log and error messages.

[Pipeline](https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/gpupgrade:refactor_createStateDir)

### CODE REVIEW

File: cli/commanders/initialize.go [hunk 1]
```diff
@@ -20,18 +20,23 @@ import (
...
+	if err != nil {
+		return xerrors.Errorf("checking if path %q exists: %w", stateDir, err)
+	}
+
+	if exists {
+		gplog.Debug("state directory %q already exists", stateDir)
 		return nil
 	}
+
+	err = os.Mkdir(stateDir, 0700)
```
Comment (dkrieger-pvtl) [inline cli/commanders/initialize.go:37]: This creates a race condition. Two callers to gpugprade initialize could both see PathExists return false, and then one of them will fail during the os.Mkdir call with a cryptic error message that that directory could not be created.

The original code does not have this issue as os.Mkdir is atomic.

**Review by dkrieger-pvtl — COMMENTED** [review]
I do not see how this refactor makes the code cleaner.  In addition, it introduces a race condition.  That race condition is unlikely, and the resolution is just to re-run `gpupgrade initialize` but I'd rather not have such things in the code.  

I would like us to solve the issue of `utils.GetStateDir()` being allowed to grab `GPUPGRADE_HOME` multiple independent times and hence picking up different env variables to think the state dir has moved, but that's a different problem than that addressed here.

File: cli/commanders/initialize.go [hunk 2]
```diff
@@ -20,18 +20,23 @@ import (
...
+// Create the state directory in the CLI to properly control the directory
+// creation and destruction.
 func CreateStateDir() (err error) {
 	stateDir := utils.GetStateDir()
-	err = os.Mkdir(stateDir, 0700)
-	if os.IsExist(err) {
-		gplog.Debug("State directory %s already present...skipping", stateDir)
+	exists, err := upgrade.PathExist(stateDir)
+	if err != nil {
+		return xerrors.Errorf("checking if path %q exists: %w", stateDir, err)
```
[the remark this answers is not in the document]
Comment (kalensk) [reply cli/commanders/initialize.go:29]: Sure, but do keep in mind the following ideas when returning errors:
- https://github.com/bahlo/go-styleguide#add-context-to-errors
- “When adding context to returned errors, keep the context succinct by avoiding phrases like "failed to", which state the obvious and pile up as the error percolates up through the stack:”
- Example adding short context: 
  - could not read config: open failed: open /Users/dfc/.settings.xml: no such file or directory
  - could not write config: write failed: input/output error
  - x: y: new store: the error

**Comment by kalensk** [comment]
> I do not see how this refactor makes the code cleaner. In addition, it introduces a race condition. That race condition is unlikely, and the resolution is just to re-run gpupgrade initialize but I'd rather not have such things in the code.

I find golangs use of inspecting the error to see if a path exists not readable. Hence, I find the PR much more easier to follow via using the `pathExists` helper.

> I would like us to solve the issue of utils.GetStateDir() being allowed to grab GPUPGRADE_HOME multiple independent times and hence picking up different env variables to think the state dir has moved, but that's a different problem than that addressed here.

As stated that is an entirely different issue and tangential to the intent of PR.

**Comment by kalensk** [comment]
Closing this PR as its more work to push through than the benefit.

**Pull request closed without merging**